lib/vdsm/storage/blockVolume.py

```diff
@@ -164,7 +164,16 @@ def getParent(self):
         """
         Return parent volume UUID
         """
-        return self.getParentTag()
+        parent_tag = self.getParentTag()
+        parent_meta = self.getParentMeta()
+        if parent_tag != parent_meta:
+            self.log.warning(
+                "volUUID=%s parent in tag (%s) and metadata (%s) differ",
+                self.volUUID, parent_tag, parent_meta)
+            lvm.reload_lvs(self.sdUUID)
+            parent_tag = self.getParentTag()
+            self.log.debug("Parent tag after reload LVs, pUUID=%s", parent_tag)
+        return parent_tag
```

Comment:
Logging debug message and returning the value without checking it is not helpful, since we know it can be wrong for some time until the volume is updated.

If the the values do not match after the refresh, we want to warn about it, so we must check again.

So now we have 3 options:
- return the non-matching tag - bad!
- return the parent meta - better, but are we sure that we want to perform operations on the volume when the volume metadata is incorrect? This can become worse.
- Fail - this ensures that the volume cannot change while the metadata is stale.

Please check how failing affects the system. You wrote about syncVolumeChain loop but I don't see the loop there. Please share logs showing the issue.

Reply:
Logs:
[vdsm.parent_raise.log](https://github.com/oVirt/vdsm/files/9165847/vdsm.parent_raise.log)

With code:
```python
    def getParent(self):
        """
        Return parent volume UUID
        """
        parent_tag = self.getParentTag()
        parent_meta = self.getParentMeta()
        if parent_tag != parent_meta:
            self.log.warning(
                "volUUID=%s parent in tag (%s) and metadata (%s) differ",
                self.volUUID, parent_tag, parent_meta)
            lvm.invalidateVG(self.sdUUID)
            parent_tag = self.getParentTag()
            self.log.debug("Parent tag after reload LVs, pUUID=%s", parent_tag)
            if parent_tag != parent_meta:
                raise se.ImageIsNotLegalChain(self.volUUID)
        return parent_tag
```

And testing VM with two disks (os + 1 data), hangs at removing the second snapshot. The only way to get out of it is by shutting down the VM.

Reply:
Thanks for sharing the log.

Trying to understand why we see in the log (since you did not provide analysis of the events):

1. Starting merge

```
2022-07-22 09:38:07,317+0200 INFO  (jsonrpc/6) [api.virt] START merge(drive={'poolID': '9227a275-ef9f-4816-81a0-506cbd7029b5', 'volumeID': 'e5bb434e-3ff1-4690-8b1c-bd0e81481a03', 'imageID': 'a6fd8b41-7173-4d8f-b59a-eb5b3b26916e', 'domainID': 'a2ad7a23-96c5-4c23-9007-c0d1892ea07d'}, baseVolUUID='d1795003-0cb6-46a7-9b07-bc4bf0de515b', topVolUUID='7b7d2a79-1979-4340-8488-a8856ac60f57', bandwidth='0', jobUUID='7a8b2475-bfaa-485b-b15e-8d8d1984edb1') from=::ffff:192.168.122.96,40192, flow_id=206af0ed-ee49-459
3-a11e-bb413367761a, vmId=595b9b5c-1779-4b84-be28-16b44e13345e (api:48)
```

2. Starting COMMIT

```
2022-07-22 09:38:07,650+0200 INFO  (jsonrpc/6) [virt.livemerge] Starting merge with job_id='7a8b2475-bfaa-485b-b15e-8d8d1984edb1', original chain=d1795003-0cb6-46a7-9b07-bc4bf0de515b < 7b7d2a79-1979-4340-8488-a8856ac60f57 < e5bb434e-3ff1-4690-8b1c-bd0e81
481a03 (top), disk='vda', base='vda[1]', top='vda[7]', bandwidth=0, flags=8 (livemerge:408)
```

This shows the chain we are merging:

```
d1795003-0cb6-46a7-9b07-bc4bf0de515b < 7b7d2a79-1979-4340-8488-a8856ac60f57 < e5bb434e-3ff1-4690-8b1c-bd0e81481a03 (top)
```

Based on baseVolUUID='d1795003-0cb6-46a7-9b07-bc4bf0de515b' and topVolUUID='7b7d2a79-1979-4340-8488-a8856ac60f57' this is internal merge, merging contents of volume 7b7d2a79-1979-4340-8488-a8856ac60f57 into volume d1795003-0cb6-46a7-9b07-bc4bf0de515b.

When the merge is done, we change the top volume parent to d1795003-0cb6-46a7-9b07-bc4bf0de515b. The parent tag will be updated when volume 7b7d2a79-1979-4340-8488-a8856ac60f57 is deleted.

The merge completed:

```
2022-07-22 09:38:07,774+0200 INFO  (libvirt/events) [virt.vm] (vmId='595b9b5c-1779-4b84-be28-16b44e13345e') Block job 7a8b2475-bfaa-485b-b15e-8d8d1984edb1 type COMMIT for drive vda has completed (vm:5969)
```

At this point qemu updated the top volume backing file to d1795003-0cb6-46a7-9b07-bc4bf0de515b, and we can see this via libvirt chain. Our metadata parent and parent tag match but both are incorrect.

Jobs switch to CLEANUP phase:

```
2022-07-22 09:38:08,621+0200 INFO  (jsonrpc/7) [virt.livemerge] Job 7a8b2475-bfaa-485b-b15e-8d8d1984edb1 switching state from COMMIT to CLEANUP (livemerge:172)
```

Syncing volume chain after merge: 

```
2022-07-22 09:38:08,677+0200 INFO  (merge/7a8b2475) [vdsm.api] START imageSyncVolumeChain(sdUUID='a2ad7a23-96c5-4c23-9007-c0d1892ea07d', imgUUID='a6fd8b41-7173-4d8f-b59a-eb5b3b26916e', volUUID='e5bb434e-3ff1-4690-8b1c-bd0e81481a03', newChain=['d1795003-0cb6-46a7-9b07-bc4bf0de515b', 'e5bb434e-3ff1-4690-8b1c-bd0e81481a03']) from=internal, task_id=4717b35d-38a0-4833-8af3-de80bbae47d8 (api:48)
```

This shows the current chain reported by libvirt:

```
newChain=['d1795003-0cb6-46a7-9b07-bc4bf0de515b', 'e5bb434e-3ff1-4690-8b1c-bd0e81481a03']
```

We see that the merged volume is not part of the chain, good.

imageSyncVolumeChain detect the state corrrectly:

```
2022-07-22 09:38:08,985+0200 INFO  (merge/7a8b2475) [storage.image] Current chain=d1795003-0cb6-46a7-9b07-bc4bf0de515b < 7b7d2a79-1979-4340-8488-a8856ac60f57 < e5bb434e-3ff1-4690-8b1c-bd0e81481a03 (top)  (image:899)
2022-07-22 09:38:09,087+0200 INFO  (merge/7a8b2475) [storage.image] Internal volume 7b7d2a79-1979-4340-8488-a8856ac60f57 removed from actual chain, linking child volume e5bb434e-3ff1-4690-8b1c-bd0e81481a03 to parent volume d1795003-0cb6-46a7-9b07-bc4bf0de515b (image:936)
```

At this point the metadata parent is correct, matching the actual chain (qcow2 backing file) and libvirt runtime info (the backing chain in libvirt xml) - good.

Syncing volume chain completed:

```
2022-07-22 09:38:09,102+0200 INFO  (merge/7a8b2475) [vdsm.api] FINISH imageSyncVolumeChain return=None from=internal, task_id=4717b35d-38a0-4833-8af3-de80bbae47d8 (api:54)
```


Here is the first issue - the cleanup thread invokes irs.getVolumeInfo():

```
2022-07-22 09:38:09,103+0200 INFO  (merge/7a8b2475) [vdsm.api] START getVolumeInfo(sdUUID='a2ad7a23-96c5-4c23-9007-c0d1892ea07d', spUUID='9227a275-ef9f-4816-81a0-506cbd7029b5', imgUUID='a6fd8b41-7173-4d8f-b59a-eb5b3b26916e', volUUID='e5bb434e-3ff1-4690-8b1c-bd0e81481a03') from=internal, task_id=c9148392-8f54-4bc9-9f84-4875a19a206b (api:48)
```

We try to get the parent tag and find that it does not match:

```
2022-07-22 09:38:09,124+0200 WARN  (merge/7a8b2475) [storage.volumemanifest] volUUID=e5bb434e-3ff1-4690-8b1c-bd0e81481a03 parent in tag (7b7d2a79-1979-4340-8488-a8856ac60f57) and metadata (d1795003-0cb6-46a7-9b07-bc4bf0de515b) differ (blockVolume:172)
```

This is expected since the tag will be updated only later after the volume is deleted, but the warning is not clear. The message should explain what we are trying to do (get parent tag), but instead it starts with key=value. This is not helpful. This message need to be something like:

    Getting volume {vol-id} parent tag: volume tag {vol-tag} does not match metadata parent {metadata-parent}, reloading volume tags

We try to reload (good):

```
2022-07-22 09:38:09,188+0200 DEBUG (merge/7a8b2475) [storage.volumemanifest] Parent tag after reload LVs, pUUID=7b7d2a79-1979-4340-8488-a8856ac60f57 (blockVolume:175)
```

As expected this does not help.

The error raised is swallowed by merge code:

```
2022-07-22 09:38:09,188+0200 DEBUG (merge/7a8b2475) [storage.volumemanifest] Failed to get volume info: Image is not a legal chain: ('e5bb434e-3ff1-4690-8b1c-bd0e81481a03',) (volume:268)
```

This is logged as debug message instead of showing the traceback. We need to fix this.

Also, the error reported is bogus:

    Image is not a legal chain: ('e5bb434e-3ff1-4690-8b1c-bd0e81481a03',)

and unhelpful. The should be a new error explaining the isssue, and showing all the 
relevant context:

    VolumeParentMismatch: volume tag {vol-tag} does not match metadata parent {metadata-parent}

So we should not reuse the InvalidChain error which is about different issue.

The next log is useless log we should remove, again without any useful message:

```
2022-07-22 09:38:09,189+0200 INFO  (merge/7a8b2475) [storage.volumemanifest] a2ad7a23-96c5-4c23-9007-c0d1892ea07d/a6fd8b41-7173-4d8f-b59a-eb5b3b26916e/e5bb434e-3ff1-4690-8b1c-bd0e81481a03 info is {'apparentsize': '0', 'truesize': '0', 'status': 'INVALID', 'children': []} (volume:282)
```

But we can see in the log that volume info is incomplete - many details are missing in the dict - and this causes the next error:

```
be28-16b44e13345e') Cleanup error for job 7a8b2475-bfaa-485b-b15e-8d8d1984edb1: 'format' (livemerge:985)
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/vdsm/virt/livemerge.py", line 970, in run
    self.vm.sync_volume_chain(self.drive)
  File "/usr/lib/python3.6/site-packages/vdsm/virt/vm.py", line 6027, in sync_volume_chain
    driveFormat = res['info']['format'].lower()
KeyError: 'format'
```

So the important error is missing (parent tag mismatch), and we complain about invalid info from the volume, caused by incorrect error handling when preparing Volume.getInfo() response.

Since the cleanup thread failed, we retry the operation, and since it will continue to fail at this situation, we are in the failure loop.

So this proves that we cannot raise the the tag does not match, which is good - so our options are:

- Report the incorrect tag as the parent
- Report the parent from the metadata

The first option does not make any sense since the tag is wrong. So we should report the metadata tag.

So I think the solution should be:

1. Check that parent tag matches metadata
2. Reload lv tags if tag does not match
3. Warn if tag does not match after reloading
4. return parent from metadata

Please try this and report the results.

Reply:
Exactly, that's what was my conclusion too, but I couldn't give (yet) a more detailed explanation.

I already did try last week returning the metadata parent, and I know it works. I will go for that solution, and check again before pushing.

But I wonder, as either both parents are either equal, or metadata is more up to date, then if we had just returned the metadata parent this bug would've never been found.

Reply:
Right after the parent metadata is updated in the logs as part of the `imagesyncVolumeData` that is executed after a live merge:

```
2022-07-25 10:14:13,679+0200 INFO  (merge/a1f89615) [storage.image] Current chain=6b4e0816-2cdc-43d0-ba5a-7753fb376150 < 76599768-1f6c-47cd-baae-db83b8056f10 < e4238fdb-7890-44c8-ace3-086550b2edec (top)  (image:899)
2022-07-25 10:14:13,746+0200 INFO  (merge/a1f89615) [storage.image] Internal volume 76599768-1f6c-47cd-baae-db83b8056f10 removed from actual chain, linking child volume e4238fdb-7890-44c8-ace3-086550b2edec to parent volume 6b4e0816-2cdc-43d0-ba5a-7753fb376150 (image:936)
```

I can see that, as expected, metadata and volume tag do not match when `getVolumeInfo` is invoked right after:
```
2022-07-25 10:14:13,842+0200 WARN  (merge/a1f89615) [storage.volumemanifest] Getting volume e4238fdb-7890-44c8-ace3-086550b2edec PUUID: volume tag 76599768-1f6c-47cd-baae-db83b8056f10 does not match metadata parent 6b4e0816-2cdc-43d0-ba5a-7753fb376150, using metadata parent (blockVolume:179)
```

And returning metadata is right. All occurrences of this warning happen always once per volume in the chain in the same point, right after the `imagesyncVolumeData` that updates the parent metadata due to a live merge.